pkg/loadbalancer/loadbalancer.go

```diff
@@ -475,17 +475,21 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
 	}
 }
 
-// Hash calculates L3n4Addr's internal SHA256Sum.
+// Hash calculates a unique string of the L3n4Addr e.g for use as a key in maps
 func (a L3n4Addr) Hash() string {
+	// Note: 32 bytes is not enough for long IPv6 addresses, but it is cheaper
+	// to reallocate on overflow than to check if a.IP is IPv6
+	b := make([]byte, 0, 32)
+
+	b = append(b, a.IP.String()...)
```

Comment:
why don't we do something like:
```
ip := a.IP.String()
if bytes.IndexRune(ip, ':') < 0 {
    b = make([]byte, 0, maxV4 /*(=32)*/)
} else {
    b = make([]byte, 0, maxV6 /*(=39? + 5 + sizeOfScope)*/)
}

Reply:
This seems to do the trick without a measurable overhead. Will change.

Reply:
Actually, after some more benchmarking, simply relying on reallocation is still faster in most cases than trying to be smart about it:

---

Fastest / Most imprecise:

``` go
const lenIPv4 = 15
const lenProto = 1
const lenPort = 6
const lenScope = 2

// Note: The capacity might not be enough for long IPv6 addresses, but it is cheaper
// to reallocate on overflow than to check the length of a.IP
b := make([]byte, 0, lenIPv4+lenProto+lenPort+lenScope)
```

```
BenchmarkL3n4Addr_Hash_IPv4-8         	16097442	        70.7 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 9431970	       125 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 5561029	       215 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-8     	 3719448	       314 ns/op
```

---

Middle ground:

``` go
const lenIPv4 = 15
const lenIPv6 = 39
const lenProto = 1
const lenPort = 6
const lenScope = 2

var b []byte
ip := a.IP.String()
if strings.IndexRune(ip, ':') < 0 {
    b = make([]byte, 0, lenIPv4+lenProto+lenPort+lenScope)
} else {
    b = make([]byte, 0, lenIPv6+lenProto+lenPort+lenScope)
}
```

```
BenchmarkL3n4Addr_Hash_IPv4-8         	14471590	        80.9 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 9139532	       132 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 6556764	       204 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-8     	 3862281	       313 ns/op
```

---

Slowest / Most-precise:

```go
const lenProto = 1
const lenPort = 6
const lenScope = 2

ip := a.IP.String()
b := make([]byte, 0, len(ip)+lenProto+lenPort+lenScope)
```

```
BenchmarkL3n4Addr_Hash_IPv4-8         	12182581	        97.3 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-8   	 7979997	       147 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-8    	 5991332	       215 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-8     	 3743278	       317 ns/op
```